### ISSUE TITLE
Docs and docstrings for io.*.seek are inconsistent

### ISSUE THREAD

**Issue by MatthiasWiesmann** [python/cpython#107801]
# Documentation

The documentation on text handles is misleading, in particular for the `seek` method.

# Steps to reproduce

```
handle = open('/tmp/lines.txt')
help(handle.seek)
```

This produces the following documentation:
```
Help on built-in function seek:

seek(cookie, whence=0, /) method of _io.TextIOWrapper instance
    Change stream position.
    
    Change the stream position to the given byte offset. The offset is
    interpreted relative to the position indicated by whence.  Values
    for whence are:
    
    * 0 -- start of stream (the default); offset should be zero or positive
    * 1 -- current stream position; offset may be negative
    * 2 -- end of stream; offset is usually negative
    
    Return the new absolute position.
```

# Issues

* The documentation talks about a byte offset, but the interface mentions a cookie. 
* The constants defined in `os` module, like `os.SEEK_END` are not mentioned.
* whence=2 with negative offset fails.

Generally, the behaviour is very inconsistent, seek relative to the end fail, seek relative to the start works, but might yield a situation when `read` fails. 

[Full investigation of the issue](https://wiesmann.codiferes.net/wordpress/archives/36371) 

### Linked PRs
* gh-107899
* gh-107933
* gh-107935
* gh-107936
* gh-108086
* gh-108087
* gh-108099
* gh-108100
* gh-108108
* gh-108136
* gh-108137
* gh-108262
* gh-108264
* gh-108265
* gh-108268
* gh-108547
* gh-108548
* gh-108655
* gh-108656

**Comment by sunmy2019** [comment]
Note: If anything is changed here, the relevant doc should also be changed.
https://docs.python.org/3/library/io.html#io.IOBase.seek

**Comment by AA-Turner** [comment]
@MatthiasWiesmann would you have time to write a PR with suggestied revisions of the help text? 

To note, Python 3.12 will include the constants in ``os`` (see #104418), though there is quite a bit of inconsistency between the various ``seek()`` methods' docstrings:

* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/iobase.c#L82-L99
* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/bufferedio.c#L1285-L1290
* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/bytesio.c#L637-L650
* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/fileio.c#L954-L969
* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/stringio.c#L464-L477
* https://github.com/python/cpython/blob/2e27da18952c6561f48dab706b5911135cedd7cf/Modules/_io/textio.c#L2428-L2433

cc @erlend-aasland as the author of #104418 if you have any thoughts.

A

**Comment by erlend-aasland** [comment]
I agree that it is confusing that the parameter names for the low-level API `seek()` aren't used consistently throughout the docs/docstrings. Luckily, these are all positional-only, so we can easily normalise them without backwards compatibility concerns.

Suggested improvements:

- Consistently name the `seek()` params
- Consistently use `io.SEEK_*` constants instead of `os.SEEK_*`
- Document `io.SEEK_` constants

**Comment by serhiy-storchaka** [comment]
1. The confusion is caused by the fact that `TextIOWrapper.seek()` has no docstring, so pydoc shows a docstring from a parent class, which is inadequate due to significant difference in semantic.
2. The first argument of `TextIOWrapper.seek()` is not a byte offset, and not an offset at all. It is a "cookie", i.e. zero or an opaque number returned by `tell()`. Passing anything other than zero and the value returned by `tell()` when `whence=SEEK_SET` has undefined behavior.
3. SEEK_CUR and SEEK_END only support zero as the first argument.

Neither _forward seeking_ nor _backward seeking_ are supported. The only supported are _rewinding to the start_ and _restoring the previous position_ saved by `tell()`. Confusion could be avoided if `TextIOWrapper.tell()` returned a special type instead of integer, but for now it would be a large breaking change (because the cookie can be saved in a file as an integer and used in other process).

**Comment by erlend-aasland** [comment]
As a data point: on macOS, Linux, and FreeBSD, the `lseek` and `fseek` manpages uses *offset* and *whence* as parameter names.

**Comment by serhiy-storchaka** [comment]
Because they only work with binary files. They do not work with TextIOWrapper which support multibyte stateful codecs.

**Comment by erlend-aasland** [comment]
There are a lot of tasks here. Suggesting to proceed in this order:

1. Use *whence* instead of *how* as the name of the third position-only parameter of `os.lseek`. Rationale: the `SEEK_*` constants are also used for this function. If the `SEEK_*` constants are to talk about the *whence* parameter, we should use it for `os.lseek` as well.
2. Improve the documentation of `os.SEEK_SET`, `os.SEEK_CUR` and `os.SEEK_END`; explain that they are used for the *whence* parameter of `os.lseek` and of the `seek()` method of file objects; add a very short description to each constant.
3. Document `os.SEEK_HOLE` and `os.SEEK_DATA` (exposed in `os` though `posix`)
4. Consistently name `_io.*.seek()` params **where is makes sense**.
5. Improve the **docs** of the `seek()` methods (Doc/library/io.rst)
6. Improve the **docstrings** of the `seek()` methods

**Comment by erlend-aasland** [comment]
> Because they only work with binary files. They do not work with TextIOWrapper which support multibyte stateful codecs.

Yes, that is fine; `TextIOWrapper` is a special case. For the other cases, it may make sense to use a consistent naming.

**Comment by erlend-aasland** [comment]
`__text_signature__` for `_io` types with `seek`:

<details>
<summary>3.11</summary>

```
BufferedRandom.__text_signature__: ($self, target, whence=0, /)
BufferedReader.__text_signature__: ($self, target, whence=0, /)
BufferedWriter.__text_signature__: ($self, target, whence=0, /)
BytesIO.__text_signature__: ($self, pos, whence=0, /)
FileIO.__text_signature__: ($self, pos, whence=0, /)
StringIO.__text_signature__: ($self, pos, whence=0, /)
TextIOWrapper.__text_signature__: ($self, cookie, whence=0, /)
```

</details>

<details>
<summary>3.12</summary>

```
BufferedRWPair.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
BufferedRandom.__text_signature__: ($self, target, whence=0, /)
BufferedReader.__text_signature__: ($self, target, whence=0, /)
BufferedWriter.__text_signature__: ($self, target, whence=0, /)
BytesIO.__text_signature__: ($self, pos, whence=0, /)
FileIO.__text_signature__: ($self, pos, whence=0, /)
StringIO.__text_signature__: ($self, pos, whence=0, /)
TextIOWrapper.__text_signature__: ($self, cookie, whence=0, /)
_BufferedIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_IOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_RawIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_TextIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
```

</details>

<details>
<summary>3.13</summary>

```
BufferedRWPair.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
BufferedRandom.__text_signature__: ($self, target, whence=0, /)
BufferedReader.__text_signature__: ($self, target, whence=0, /)
BufferedWriter.__text_signature__: ($self, target, whence=0, /)
BytesIO.__text_signature__: ($self, pos, whence=0, /)
FileIO.__text_signature__: ($self, pos, whence=0, /)
StringIO.__text_signature__: ($self, pos, whence=0, /)
TextIOWrapper.__text_signature__: ($self, cookie, whence=0, /)
_BufferedIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_IOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_RawIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
_TextIOBase.__text_signature__: ($self, offset, whence=os.SEEK_SET, /)
```

</details>

**Comment by serhiy-storchaka** [comment]
Even in C, `fseek`/`ftell` and `lseek` and not interoperable. You will have issues when pass the result of `lseek` to `fseek` on Windows (because newline translation and buffering). There were such issues in the CPython interpreter. So "offset" in `fseek`/`ftell` is not just an offset.

`fgetpos` and `fsetpos` are more modern analogs of `fseek`/`ftell`, and they work with position which is not a number.

**Comment by serhiy-storchaka** [comment]
No, it was response to https://github.com/python/cpython/issues/107801#issuecomment-1676868944.

As for your proposition, it all looks reasonable. The main entries are 6 and 5. Perhaps there should be only two separate docstrings: for binary and text streams, all other classes perhaps can inherit them from parent. The module documentation can be more verbose and provide more specific details. Other entries can go to other issues.

**Comment by hugovk** [comment]
Lots of merged PRs here. 🚀 

Anything still to do, or can the issue be closed?

**Comment by vadmium** [comment]
Issue #82969 is open about IOBase.seek/tell vs TextIOWrapper and the offset/cookie parameter
Issue #122299 is about the byte vs character seek offsets

**Comment by picnixz** [comment]
I'll close this issue as `completed` since #82969 was closed and #122299 is likely to be closed as well IMO. We can always add more docs if needed and use that issue as a tracker.
